### ISSUE TITLE
NMS: separate results per class

### ISSUE THREAD

**Issue by albertfaromatics** [zzh8829/yolov3-tf2#271]
According to the TF docs (https://www.tensorflow.org/api_docs/python/tf/image/combined_non_max_suppression), in order to get results of NMS per class, the input shape of the boxes tensor should be [batch_size, num_boxes, q, 4], where q is equal to the number of classes.

Here we have q=1, so we are applying the NMS to all the boxes, without splitting per class.

How can I make it so the results are per class? My images can have overlapping boxes of different classes, and I think that they sometimes are being filtered out by the NMS

**Comment by Morgensol** [comment]
If you are worried that you are filtering out overlapping boxes you can fiddle with the IOU threshold you have so that it wont filter out, that is a quick way of checking if your suspicion is correct.
Though i think that it is more likely that it has something to do with your anchorboxes (only if you are using custom classes though).

**Comment by albertfaromatics** [comment]
The IOU threshold I think it will not solve the problem. I can have a full "small" bounding box inside another "big" bounding box, like a person inside a big van (I'm not on people nor cars, but as example).

I'm think that sometimes the "small" boxes are being filtered out, because the IOU is 1, as the small is totally inside the big box.

Maybe it is a YOLO limitation and I'm trying to fight against it

**Comment by Morgensol** [comment]
No it is definetly possible to detect multiple objects on top of each other, that is primarily what the anchorboxes are there for, for example; if you have a person infront of a car one box in the grid will detect a long horizontal rectangular object and give you a positive, and another vertical rectangular object will detect the person. So it should def be possible. 

What you can do is just set the IOU threshold such that it allows a 100% overlap.

From my personal experience i have no issue in detecting multiple objects on top of each other (though usually that is an false detection).

**Comment by albertfaromatics** [comment]
But if I set IOU threshold to 1, I will allow all bounding boxes to pass, getting hundreds of them, and making the NMS useless but a filter per score, not IOU, am I right?

That's why I want to make the NMS per class, not global, so each class has its own result and overlapping boxes of different classes are not taken into account when computing IOU between boxes.

**Comment by Morgensol** [comment]
Yes, it is definitely not a good permanent solution, but if you do this, you should see some detections on the faces, and yes possibly in the 100's, but it will confirm you theory, and thus means that you are correct on your assumption about the NMS.

I will say though, it doesnt sound crazy what you are saying, it is quite possible that you need to change the NMS to do just that, unless i have misunderstood it aswell.

**Comment by albertfaromatics** [comment]
Good point. I'll try to set IOU to 1, and see how many boxes I get and if I get overlapping results of different classes.

Then I'll try to research how to convert the input parameters of combined_nms to allow distinction between classes. The TF docs are not clear at all though...

Thanks @Morgensol
